### PR TITLE
GAWB-2958: Orch endpoint for user agreement 

### PR DESCRIPTION
**Pull request by kyuksel**
This PR is not ready to merge but was opened to solicit early feedback. The functionality is there but it needs tests, and possibly some refactoring. 

It persists and keeps track of free trial user's agreement to Onix' terms. The highlights are as follows:

- A new `userAgreed` field was added to Thurloe.
- The flag is set to `false` at user enabling time. **Note** that if the user was already enabled and is attempted to be enabled again, `userAgreed` flag is **not** set to `false`. I can't see why this could be a problem but wanted to point out.
- The flag is set to `true` when there is an authorized `PUT` to `/api/trial/userAgreement`
- At user enrollment time, the flag is checked. If it was not set to `true`, the user is barred from enrolling. If it was, everything proceeds as before.

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Review by davidangb — COMMENTED** [review]
I did not review in any depth, but architecturally/design-wise this looks right!

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala [hunk 1]
```diff
@@ -187,17 +189,21 @@ final class TrialService
...
-              val expirationDate = now.plus(FireCloudConfig.Trial.durationDays, ChronoUnit.DAYS)
-              val enrolledStatus = status.copy(
-                state = Some(TrialStates.Enrolled),
-                enrolledDate = now,
-                expirationDate = expirationDate
-              )
-              thurloeDao.saveTrialStatus(userInfo.id, userInfo, enrolledStatus) map { _ =>
-                // TODO: add user to free-trial billing project / create said project if necessary
-                RequestComplete(NoContent)
+              if (status.userAgreed) {
```
Comment (aednichols) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala:195]: We will have a mild merge conflict with `an-GAWB-2910`

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala [hunk 2]
```diff
@@ -253,4 +259,16 @@ final class TrialService
...
     Future(RequestComplete(OK, trialDAO.projectReport))
 
+  private def recordUserAgreement(userInfo: UserInfo): Future[PerRequestMessage] = {
+    thurloeDao.getTrialStatus(userInfo.id, userInfo) flatMap {
+      case Some(status) => status.state match {
+        case Some(TrialStates.Enabled) => thurloeDao.saveTrialStatus(userInfo.id, userInfo, status.copy(userAgreed = true)) flatMap {
+          case Success(_) => Future(RequestComplete(NoContent))
+          case Failure(ex) => Future(RequestComplete(InternalServerError, ex.getMessage))
+        }
+        case _ => Future(RequestCompleteWithErrorReport(BadRequest, "You are not Enabled to agree to terms."))
```
Comment (aednichols) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala]: I think we should keep things like their trial status (`Enabled`) opaque from end users, it's not like there's anything they can do about it

File: src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala [hunk 3]
```diff
@@ -20,37 +20,51 @@ trait TrialApiService extends HttpService with PerRequestCreator with FireCloudD
...
+            requireUserInfo() { managerInfo => // We will need the manager's credentials for authentication downstream
+              entity(as[Seq[String]]) { userEmails =>
+                requestContext =>
+                  perRequest(requestContext,
+                    TrialService.props(trialServiceConstructor),
+                    updateUsers(managerInfo, TrialOperations.withName(operation), userEmails))
+              }
+            }
+          } ~
+            path("projects") {
```
Comment (aednichols) [inline src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala]: Nit: the block `48`-`67` is indented one level too far

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala (hunk 2 above)
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala]: I wanted to distinguish between the error cases for developer purposes but I agree with your statement.

File: src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala (hunk 3 above)
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala]: Thanks for the catch!

File: src/main/scala/org/broadinstitute/dsde/firecloud/webservice/UserApiService.scala [hunk 4]
```diff
@@ -129,12 +129,25 @@ trait UserApiService extends HttpService with PerRequestCreator with FireCloudRe
...
-        post {
-          requireUserInfo() { userInfo => requestContext =>
-            perRequest(requestContext, TrialService.props(trialServiceConstructor),
-              TrialService.EnrollUser(userInfo)
-            )
+      pathPrefix("profile" / "trial") {
+        pathEnd {
+          post {
+            requireUserInfo() { userInfo =>
+              requestContext =>
```
Comment (ahaessly) [inline src/main/scala/org/broadinstitute/dsde/firecloud/webservice/UserApiService.scala]: nit: requestContext can be moved to line above (similar to usage in other files)
same with userAgreement below

File: src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala [hunk 5]
```diff
@@ -36,18 +37,19 @@ trait TrialApiService extends HttpService with PerRequestCreator with FireCloudD
...
-                  case "count" => Some(TrialService.CountProjects(userInfo))
-                  case "report" => Some(TrialService.Report(userInfo))
-                  case _ => None
-                }
-                if (message.nonEmpty)
-                  perRequest(requestContext, TrialService.props(trialServiceConstructor), message.get)
-                else
-                  requestContext.complete(BadRequest, ErrorReport(s"invalid operation '$op'"))
+              requireUserInfo() { userInfo =>
+                requestContext =>
```
Comment (ahaessly) [inline src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala]: nit: move requestContext back to line above. same with endpoint above

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala [hunk 6]
```diff
@@ -256,4 +262,19 @@ final class TrialService
   private def projectReport: Future[PerRequestMessage] =
     Future(RequestComplete(OK, trialDAO.projectReport))
 
-}
\ No newline at end of file
+  private def recordUserAgreement(userInfo: UserInfo): Future[PerRequestMessage] = {
+    // Thurloe errors are handled by the caller of this method
+    thurloeDao.getTrialStatus(userInfo.id, userInfo) flatMap {
+      case Some(status) =>
+        status.state match {
```
Comment (ahaessly) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala:269]: nit: can move this to line above

**Review by ahaessly — APPROVED** [review]
just a couple of formatting comments

**Review by zarsky-broad — APPROVED** [review]
Verified working with correct UI branch.

**Pull request closed without merging**